contrib/educational_decoder/zstd_decompress.c

```diff
@@ -0,0 +1,2096 @@
+    }
+    bits[num_symbs] =
+        max_bits + 1 - last_weight; // last weight is always non-zero
+
+    HUF_init_dtable(table, bits, num_symbs + 1);
+}
+
+static void HUF_free_dtable(HUF_dtable *dtable) {
+    free(dtable->symbols);
+    free(dtable->num_bits);
+    memset(dtable, 0, sizeof(HUF_dtable));
+}
+
+static void HUF_copy_dtable(HUF_dtable *dst, const HUF_dtable *src) {
+    if (src->max_bits == 0) {
+        memset(dst, 0, sizeof(HUF_dtable));
+        return;
+    }
+
+    size_t size = (size_t)1 << src->max_bits;
+    dst->max_bits = src->max_bits;
+
+    dst->symbols = malloc(size);
+    dst->num_bits = malloc(size);
+    if (!dst->symbols || !dst->num_bits) {
+        BAD_ALLOC();
+    }
+
+    memcpy(dst->symbols, src->symbols, size);
+    memcpy(dst->num_bits, src->num_bits, size);
+}
+/******* END HUFFMAN PRIMITIVES ***********************************************/
+
+/******* FSE PRIMITIVES *******************************************************/
+static inline u8 FSE_peek_symbol(FSE_dtable *dtable, u16 state) {
+    return dtable->symbols[state];
+}
+
+static inline void FSE_update_state(FSE_dtable *dtable, u16 *state,
+                                    const u8 *src, i64 *offset) {
+    const u8 bits = dtable->num_bits[*state];
+    const u16 rest = STREAM_read_bits(src, bits, offset);
+    *state = dtable->new_state_base[*state] + rest;
+}
+
+// Decodes a single FSE symbol and updates the offset
+static inline u8 FSE_decode_symbol(FSE_dtable *dtable, u16 *state,
+                                   const u8 *src, i64 *offset) {
+    const u8 symb = FSE_peek_symbol(dtable, *state);
+    FSE_update_state(dtable, state, src, offset);
+    return symb;
+}
+
+static inline void FSE_init_state(FSE_dtable *dtable, u16 *state, const u8 *src,
+                                  i64 *offset) {
+    const u8 bits = dtable->accuracy_log;
+    *state = STREAM_read_bits(src, bits, offset);
+}
+
+static size_t FSE_decompress_interleaved2(FSE_dtable *dtable, u8 *dst,
+                                          size_t dst_len, const u8 *src,
+                                          size_t src_len) {
+    if (src_len == 0) {
+        INP_SIZE();
+    }
+
+    u8 *dst_max = dst + dst_len;
+    u8 *const odst = dst;
+
+    // Find the last 1 bit
+    int padding = 8 - log2inf(src[src_len - 1]);
+
+    i64 offset = src_len * 8 - padding;
+
+    u16 state1, state2;
+    FSE_init_state(dtable, &state1, src, &offset);
+    FSE_init_state(dtable, &state2, src, &offset);
+
+    // Decode until we overflow the stream
+    // Since we decode in reverse order, overflowing the stream is offset going
+    // negative
+    while (1) {
+        if (dst > dst_max - 2) {
+            OUT_SIZE();
+        }
+        *dst++ = FSE_decode_symbol(dtable, &state1, src, &offset);
+        if (offset < 0) {
+            // There's still a symbol to decode in state2
+            *dst++ = FSE_decode_symbol(dtable, &state2, src, &offse
```

Comment:
Always check the result of `malloc()`.
You need to have a way to express "init failed".

As a simpler educational decoder, it's possible to simplify this issue by simply aborting the program
(`exit(errorNb)`).
A more "production-level" decoder will have to return some kind of "error code" that the library user can trap and use.